pkg/controller/sparkapplication/controller.go

```diff
@@ -17,465 +17,506 @@ limitations under the License.
-	kubeClient        clientset.Interface
-	queue             workqueue.RateLimitingInterface
-	cacheSynced       cache.InformerSynced
-	recorder          record.EventRecorder
-	metrics           *sparkAppMetrics
-	applicationLister crdlisters.SparkApplicationLister
-	podLister         v1.PodLister
-	ingressURLFormat  string
-	batchSchedulerMgr *batchscheduler.SchedulerManager
-}
-
-// NewController creates a new Controller.
-func NewController(
-	crdClient crdclientset.Interface,
-	kubeClient clientset.Interface,
-	crdInformerFactory crdinformers.SharedInformerFactory,
-	podInformerFactory informers.SharedInformerFactory,
-	metricsConfig *util.MetricConfig,
-	namespace string,
-	ingressURLFormat string,
-	batchSchedulerMgr *batchscheduler.SchedulerManager) *Controller {
-	crdscheme.AddToScheme(scheme.Scheme)
-
-	eventBroadcaster := record.NewBroadcaster()
-	eventBroadcaster.StartLogging(glog.V(2).Infof)
-	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{
-		Interface: kubeClient.CoreV1().Events(namespace),
-	})
-	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, apiv1.EventSource{Component: "spark-operator"})
-
-	return newSparkApplicationController(crdClient, kubeClient, crdInformerFactory, podInformerFactory, recorder, metricsConfig, ingressURLFormat, batchSchedulerMgr)
+// Add creates a new SparkApplication Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, metricsConfig *util.MetricConfig) error {
+	return add(mgr, newReconciler(mgr, metricsConfig))
 }
 
-func newSparkApplicationController(
-	crdClient crdclientset.Interface,
-	kubeClient clientset.Interface,
-	crdInformerFactory crdinformers.SharedInformerFactory,
-	podInformerFactory informers.SharedInformerFactory,
-	eventRecorder record.EventRecorder,
-	metricsConfig *util.MetricConfig,
-	ingressURLFormat string,
-	batchSchedulerMgr *batchscheduler.SchedulerManager) *Controller {
-	queue := workqueue.NewNamedRateLimitingQueue(&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(queueTokenRefillRate), queueTokenBucketSize)},
-		"spark-application-controller")
-
-	controller := &Controller{
-		crdClient:         crdClient,
-		kubeClient:        kubeClient,
-		recorder:          eventRecorder,
-		queue:             queue,
-		ingressURLFormat:  ingressURLFormat,
-		batchSchedulerMgr: batchSchedulerMgr,
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	//p := predicate.Funcs{
+	//	UpdateFunc: func(e event.UpdateEvent) bool {
+	//
+	//		oldObject := e.ObjectOld.(*v1beta1.SparkApplication)
+	//		newObject := e.ObjectNew.(*v1beta1.SparkApplication)
+	//		if !reflect.DeepEqual (oldObject.Status, newObject.Status) {
+	//			// NO enqueue request
+	//			return false
+	//		}
+	//		// ENQUEUE request
+	//		return true
+	//	},
+	//}
```

Comment:
@yuchaoran2011 ideally we should progress based only on pod transition (real resources) and not status updates. Check here: https://github.com/kubernetes-sigs/kubebuilder/issues/618 on the other idempotency could be more desired. Haven't decided on this yet. Ideally I shouldnt get any changes that I dont want to, but I think I noticed some updated that is why I have this map `sm` check next to do a proper comparison.

Reply:
Downside of keeping state is that on operator restarts we need to restore it. I havent done this yet.

Reply:
I will check if filtering out based on the request object can happen here.